File: javaDemo/src/main/java/ReadFilesDemo.java
import reactor.core.publisher.Flux;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.stream.BaseStream;

public class ReadFilesDemo {

    private static Flux<String> getLinesOfFile(String path) {
        return Flux.using(
                () -> Files.lines(Paths.get(path)),
                Flux::fromStream,
                BaseStream::close
        );
    }

    private static Flux<String> glueFiles(String filename1, String filename2, String filename3) {
        return getLinesOfFile(filename1).flatMap(lineFromFirstFile ->
                getLinesOfFile(filename2)
                        .filter(line -> line.equals(lineFromFirstFile))
                        .flatMap(lineFromSecondFile ->
                            getLinesOfFile(filename3)
                                .filter(line -> line.equals(lineFromSecondFile))
                                .map(lineFromThirdFile ->
                                    lineFromThirdFile
                            )
                )
        );
    }

    public static void main(String[] args) {
        glueFiles("first.txt", "second.txt", "third.txt")
                .subscribe(System.out::println);
    }

}

File: scalaDemo/src/main/scala/ReadFilesDemo.scala
import cats.effect._
import cats.syntax.functor._
import fs2.{Stream, io, text}
import java.nio.file.Paths
import java.util.concurrent.Executors

import scala.concurrent.{ExecutionContext, ExecutionContextExecutorService}

object ReadFilesDemo extends IOApp {

  implicit val blockingExecutionContext: ExecutionContextExecutorService =
    ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(2))

  def readFile(filename: String)
              (implicit blockingExecutionContext: ExecutionContextExecutorService): Stream[IO, String] =
    io
      .file
      .readAll[IO](Paths.get(filename), blockingExecutionContext, 4096)
      .through(text.utf8Decode)
      .through(text.lines)

  def findUniqueLines(filename1: String, filename2: String, filename3: String): Stream[IO, String] =
    for {
      lineFromFirstFile <- readFile(filename1)
      lineFromSecondFile <- readFile(filename2).filter(_.equals(lineFromFirstFile))
      result <- readFile(filename3).filter(_.equals(lineFromSecondFile))
    } yield result

  def program(glueFiles: Stream[IO, String])
             (implicit blockingExecutionContext: ExecutionContextExecutorService): IO[Unit] =
      glueFiles
      .evalMap(x => IO(println(x)))
      .compile
      .drain
      .guarantee(IO(blockingExecutionContext.shutdown()))

  override def run(args: List[String]): IO[ExitCode] =
    program(findUniqueLines("first.txt", "second.txt", "third.txt")).as(ExitCode.Success)
}
